### PR TITLE
T4rec refactor(Part 1)

### PR DESCRIPTION
**Pull request by nzarif**
### Goals :soccer:
The final goal of these changes is to:
1. Be able to use HF Trainer for training all types of tasks (i.e. `BinaryClassificationTask` and `RegressionTask` as well as `NextItemPredictionTask`) to leverage the multi-gpu support already implemented in HF Trainer.
2. Standardize inference to help with integration with Triton.

### Implementation Details :construction:
- Standardize the output of `forward()` functions within T4Rec. All Tasks, Heads and Models will return:
1. A dictionary of `{"loss":Tensor, "labels": Union[Tensor,Dict{"task_name":Tensor}], "predictions":Union[Tensor,Dict{"task_name":Tensor}]}` during **training** and **evaluation**
3. A Tensor during **inference**
To learn more, refer to [this diagram](https://excalidraw.com/#json=Ahjez-KBfH64FYU0aBLfd,Wg7fQq2OiqL4I0kZUKSh-Q).
- Refactor T4Rec flags used within `forward()` functions of tasks, Heads and Models:
1. Removed `hf_format` flag as it was not clear and not needed after standardizing the output format.
2. Added `testing` and `training` flags to indicate whether we are doing training, evaluation or inference. To know how these flags work refer to top half of [this diagram](https://excalidraw.com/#json=Ahjez-KBfH64FYU0aBLfd,Wg7fQq2OiqL4I0kZUKSh-Q).
3. Removed `ignore_masking` flag because knowing the mode (train, eval or inference) with the help of `training` and `testing` flags means we can know if masking must be applied or ignored.
- `targets` was added as an input argument to `forward()` functions that will be called for `BinaryClassificationTask` and `RegressionTask`. So from now on, we can call `forward()` from HF Trainer class to train these tasks and there is no more need to call custom `compute_loss()` function.

### Testing Details :mag:
Go to [CI integration tests](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/ci/test_integration.sh) and run the tests included there using the dataset already downloaded and unzipped by the shell script.

### CODE REVIEW

**Review by sararb — CHANGES_REQUESTED** [review]
Thank you for refactoring the T4Rec outputs! It will simplify the API usage and unify the training, evaluation, and prediction modes for all prediction tasks. ☺️

I requested some changes to the code to ensure the correct usage of the different modes. I also think this work is an excellent opportunity to add/update docstrings for the different refactored output methods (forward, calculate_metrics, prediction_tasks, and head). But it could be done later, once the code base is finalized.

File: transformers4rec/torch/model/base.py [hunk 1]
```diff
@@ -193,22 +205,24 @@ def compute_loss(
...
+        testing: bool = True,
         **kwargs,
     ) -> Dict[str, torch.Tensor]:
         if isinstance(targets, dict) and self.target_name:
             targets = targets[self.target_name]
 
         outputs = {}
         if forward:
-            predictions = self(predictions, ignore_masking=ignore_masking)
+            predictions = self(predictions, training=training, testing=testing)
```
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/model/base.py:225]: Why do we need the labels for? I can not see labels being used right now in the `calculate_metrics` function.

File: transformers4rec/torch/model/base.py [hunk 2]
```diff
@@ -534,40 +543,63 @@ def __init__(
...
+            outputs = {"loss": loss, "labels": labels, "predictions": predictions}
 
-        loss_tensor = torch.stack(losses)
+        else:
+            for head in self.heads:
+                outputs.update(
+                    head(inputs, call_body=True, training=training, testing=testing, **kwargs)
+                )
+                if len(outputs) == 1:
+                    return outputs[list(outputs.keys())[0]]
```
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/model/base.py:576]: Changed this to `return outputs[list(outputs.values())[0]]`

File: transformers4rec/torch/trainer.py [hunk 3]
```diff
@@ -394,11 +395,6 @@ def evaluation_loop(
             else self.args.prediction_loss_only
         )
 
-        if description == "Prediction":
```
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/trainer.py:397]: training should not be modified and must remain as is?

[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/trainer.py:397]: Why `testing=Flase if description=="prediction"`? Shouldn't it be the other way?

File: transformers4rec/torch/model/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sararb) [reply transformers4rec/torch/model/base.py:225]: The `targets` is used to compute the metrics in line 233. For the tasks that modify/update the targets (as in next-item prediction task): the targets tensor should be updated based on the value of `predictions['label']`, returned by the forward call

File: transformers4rec/torch/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sararb) [reply transformers4rec/torch/trainer.py:397]: To have the same logic as in Merlin Models, the flag `testing==True` is reserved for evaluation mode. In HF, the prediction mode (i.e test or inference) is defined by `description ==Prediction` while evaluation is defined by `description==evaluation`.

File: transformers4rec/torch/model/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/model/base.py:225]: I don't exactly understand this part. Maybe I'd better push the new version I have then we can discuss on that version. Right now line 233 points to something outside of `calculate_metrics`

File: transformers4rec/torch/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/trainer.py:397]: I see, thanks!

[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/trainer.py:397]: But we assumed `testing=True` shows testing, not evaluation.

[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/trainer.py:397]: Let's say I add this here:
`if description=="prediction":
    testing=False
    training=False`
and
`if description=="evaluation":
    testing=True
    training=False`
  Is this okay?

File: transformers4rec/torch/model/base.py [hunk 4]
```diff
@@ -543,52 +523,64 @@ def __init__(
...
+                predictions = list(predictions.values())[0]
+            outputs = {"loss": loss, "labels": labels, "predictions": predictions}
 
-        loss_tensor = torch.stack(losses)
+        else:
+            for head in self.heads:
+                outputs.update(
+                    head(inputs, call_body=True, training=training, testing=testing, **kwargs)
+                )
+                if len(outputs) == 1:
```
Comment (sararb) [inline transformers4rec/torch/model/base.py:556]: This condition should happen outside of the for loop otherwise we will always return the outputs of the first head.

File: transformers4rec/torch/model/base.py [hunk 5]
```diff
@@ -619,7 +611,7 @@ def forward(self, inputs, *args, **kwargs):
                 return self.parent(inputs, *args, **kwargs)
 
             def training_step(self, batch, batch_idx):
-                loss = self.parent.compute_loss(*batch)
+                loss = self.parent(*batch, training=True, testing=False)
```
Comment (sararb) [inline transformers4rec/torch/model/base.py:614]: do we need to specify the key `loss` here? 
```suggestion
                loss = self.parent(*batch, training=True, testing=False)['loss']
```

File: transformers4rec/torch/trainer.py [hunk 6]
```diff
@@ -459,7 +461,8 @@ def evaluation_loop(
                 inputs,
                 prediction_loss_only,
                 ignore_keys=ignore_keys,
-                ignore_masking=ignore_masking,
+                training=False,
+                testing=True,
```
Comment (sararb) [inline transformers4rec/torch/trainer.py:465]: the testing flag should be set based on the value defined in the `if description=="prediction":` statement. 
```suggestion
                testing=testing,
```

**Review by sararb — CHANGES_REQUESTED** [review]
Thank you for the updates! I have just left three minor changes.

File: transformers4rec/torch/model/base.py (hunk 5 above)
Comment (nzarif) [reply transformers4rec/torch/model/base.py:614]: Yes, changed it.

**Comment by sararb** [comment]
Closing this PR as this work was split into three PRs (#542, #544, and #554)
